x/protorev/keeper/epoch_hook.go

```diff
@@ -0,0 +1,117 @@
+}
+
+// Struct used to track the pool with the highest liquidity
+type LiquidityPoolStruct struct {
+	Liquidity sdk.Int
+	PoolId    uint64
+}
+
+var (
+	_ epochstypes.EpochHooks = EpochHooks{}
+)
+
+func (k Keeper) EpochHooks() epochstypes.EpochHooks {
+	return EpochHooks{k}
+}
+
+// BeforeEpochStart is the epoch start hook.
+func (h EpochHooks) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	return nil
+}
+
+// AfterEpochEnd is the epoch end hook. The module will update all of the pools in the store that are
+// used for trading.
+func (h EpochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	switch epochIdentifier {
+	case "week":
+		// Reset the pools in the store
+		h.k.DeleteAllAtomPools(ctx)
+		h.k.DeleteAllOsmoPools(ctx)
+
+		// Update the pools in the store
+		return h.k.UpdatePools(ctx)
+	}
+	return nil
+}
+
+func (k Keeper) UpdatePools(ctx sdk.Context) error {
+	// Get the highest liquidity pools
+	osmoPools, atomPools, err := k.GetHighestLiquidityPools(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Update the pools in the store
+	for token, poolInfo := range osmoPools {
+		k.SetOsmoPool(ctx, token, poolInfo.PoolId)
+	}
+	for token, poolInfo := range atomPools {
+		k.SetAtomPool(ctx, token, poolInfo.PoolId)
+	}
+
+	return nil
+}
+
+// GetHighestLiquidityPools returns the highest liquidity pools for pools that have Osmo or Atom
+// and Osmo/Atom
+func (k Keeper) GetHighestLiquidityPools(ctx sdk.Context) (map[string]LiquidityPoolStruct, map[string]LiquidityPoolStruct, error) {
+	// Get all pools
+	pools, err := k.gammKeeper.GetPoolsAndPoke(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	osmoPools := make(map[string]LiquidityPoolStruct)
+	atomPools := make(map[string]LiquidityPoolStruct)
+
+	// Iterate through all pools and find valid matches
+	for _, pool := range pools {
+		coins := pool.GetTotalPoolLiquidity(ctx)
+
+		// Pool must be active and the number of coins must be 2
+		if pool.IsActive(ctx) && len(coins) == 2 {
+			tokenA := coins[0]
+			tokenB := coins[1]
+
+			newPool := LiquidityPoolStruct{
+				PoolId:    pool.GetId(),
+				Liquidity: tokenA.Amount.Mul(tokenB.Amount),
+			}
+
+			// Check if there is a match with osmo
+			if otherDenom, match := types.CheckMatchAndReturnOther(tokenA.Denom, tokenB.Denom, types.OsmosisDenomination); match {
+				if currPool, ok := osmoPools[otherDenom]; !ok {
+					osmoPools[otherDenom] = newPool
+				} else {
+					if newPool.Liquidity.GT(currPool.Liquidity) {
+						osmoPools[otherDenom] = newPool
+					}
+				}
+			}
+
+			// Check if there is a match with atom
+			if otherDenom, match := types.CheckMatchAndReturnOther(tokenA.Denom, tokenB.Denom, types.AtomDenomination); match {
+				if currPool, ok := atomPools[otherDenom]; !ok {
+					atomPools[otherDenom] = newPool
+				} else {
+					if newPool.Liquidity.GT(currPool.Liquidity) {
+						atomPools[otherDenom] = newPool
+					}
+				}
+			}
+		}
+	}
```

Comment:
I think we can make this function shorter by just adding an extra function like `checkOsmoAtomDenomMatch`

Reply:
Are you talking about the function name or what it returns? We need some functionality to check whether any of the assets in a 2 asset pool are atom or osmo and then return the other denom in that pool that is neither atom nor osmo (used here and in `epoch_hook.go`).

Reply:
I was suggesting `checkOsmoAtomDenomMatch` as function name, but we can use a helper function here for logic and reduce code length

Reply:
Oh I see. I agree, I'll make change.